calypso/calypso.go

```diff
@@ -0,0 +1,264 @@
+package calypso
+
+import (
+	"bytes"
+
+	"go.dedis.ch/dela/crypto"
+	"go.dedis.ch/dela/serde"
+
+	"go.dedis.ch/dela/calypso/json"
+	"go.dedis.ch/dela/calypso/storage"
+	"go.dedis.ch/dela/calypso/storage/inmemory"
+	"go.dedis.ch/dela/dkg"
+	"go.dedis.ch/dela/ledger/arc"
+	"go.dedis.ch/dela/ledger/arc/darc"
+	serdej "go.dedis.ch/dela/serde/json"
+	"go.dedis.ch/kyber/v3"
+	"go.dedis.ch/kyber/v3/suites"
+	"golang.org/x/xerrors"
+)
+
+const (
+	// ArcRuleUpdate defines the rule to update the arc. This rule must be set
+	// at the write creation to allow the arc to be latter updated.
+	ArcRuleUpdate = "calypso_update"
+	// ArcRuleRead defines the arc rule to read a value
+	ArcRuleRead = "calypso_read"
+)
+
+// Suite is the Kyber suite for Pedersen.
+var suite = suites.MustFind("Ed25519")
+
+// NewCalypso creates a new Calypso
+func NewCalypso(dkg dkg.DKG) *Caly {
+	return &Caly{
+		dkg:     dkg,
+		storage: inmemory.NewInMemory(),
+	}
+}
+
+// Caly is a wrapper around DKG to provide a storage and authorization layer
+//
+// implements calypso.Secret
+type Caly struct {
+	dkg      dkg.DKG
+	dkgActor dkg.Actor
+	storage  storage.KeyValue
+}
+
+// Listen implements calypso.Secret
+func (c *Caly) Listen() error {
+	if c.dkgActor != nil {
+		return xerrors.Errorf("listen has already been called")
+	}
+
+	actor, err := c.dkg.Listen()
+	if err != nil {
+		return xerrors.Errorf("failed to listen dkg: %v", err)
+	}
+
+	c.dkgActor = actor
+
+	return nil
+}
+
+// Setup implements calypso.Secret
+func (c *Caly) Setup(ca crypto.CollectiveAuthority,
+	threshold int) (pubKey kyber.Point, err error) {
+
+	if c.dkgActor == nil {
+		return nil, xerrors.Errorf("dkg actor is nil, did you call Listen() first?")
+	}
+
+	pubKey, err = c.dkgActor.Setup(ca, threshold)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to setup dkg: %v", err)
+	}
+
+	return pubKey, nil
+}
+
+// GetPublicKey implements calypso.Secret
+func (c *Caly) GetPublicKey() (kyber.Point, error) {
+	if c.dkgActor == nil {
+		return nil, xerrors.Errorf("listen has not already been called")
+	}
+
+	pubKey, err := c.dkgActor.GetPublicKey()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get the DKG public key: %v", err)
+	}
+
+	return pubKey, nil
+}
+
+// Write implements calypso.Secret
+func (c *Caly) Write(em EncryptedMessage,
+	ac arc.AccessControl) ([]byte, error) {
+
+	var buf bytes.Buffer
+
+	_, err := em.GetK().MarshalTo(&buf)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to marshal K: %v", err)
+	}
+
+	_, err = em.GetC().MarshalTo(&buf)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to marshal C: %v", err)
+	}
+
+	key := crypto.NewSha256Factory().New().Sum(buf.Bytes())
```

Comment:
`Sum` does not do what you expect from it, for instance:
```go
func (d *digest) Sum(in []byte) []byte {
	// Make a copy of d so that caller can keep writing and summing.
	d0 := *d
	hash := d0.checkSum()
	if d0.is224 {
		return append(in, hash[:Size224]...)
	}
	return append(in, hash[:]...)

}
```